Add support for Capella

ethDreamer opened the pull request.

realbigsean committed:
add/update types

realbigsean committed:
cargo fix

realbigsean committed:
add shanghai fork version and epoch

realbigsean committed:
sync tx blobs

realbigsean committed:
some more sync boilerplate

realbigsean committed:
Some gossip work

realbigsean committed:
add `engine_getBlobV1`

realbigsean committed:
add new beacon state variant for shanghai

realbigsean committed:
some block building updates

realbigsean committed:
fix compilation errors, rename capella -> shanghai, cleanup some rebase issues

realbigsean committed:
cargo fix and some test fixes

dknopik committed:
start adding types

dknopik committed:
fix some bugx, adjust stucts

dknopik committed:
it compiles!

MariusVanDerWijden committed:
beacon: consensus: implement engine api getBlobs

dknopik committed:
implement From<FullPayload> for Payload

MariusVanDerWijden committed:
beacon: consensus: implement engine api getBlobs

dknopik committed:
progress on gossip stuff

MariusVanDerWijden committed:
network: blobs by range message

dknopik committed:
gossip boilerplate

dknopik committed:
Merge remote-tracking branch 'fork/eip4844' into eip4844

MariusVanDerWijden committed:
network stuff

dknopik committed:
storable blobs

MariusVanDerWijden committed:
network stuff

MariusVanDerWijden committed:
more network stuff

MariusVanDerWijden committed:
implement handle blobs by range req

dknopik committed:
add network config

dknopik committed:
Merge branch 'eip4844' of github.com:dknopik/lighthouse into eip4844

dknopik committed:
forgor something

MariusVanDerWijden committed:
new round of hacks (config etc)

MariusVanDerWijden committed:
more enr

MariusVanDerWijden committed:
correct fork version

MariusVanDerWijden committed:
hacky hacks

MariusVanDerWijden committed:
correct fork version

MariusVanDerWijden committed:
enable 4844 at epoch 3

realbigsean committed:
compile

realbigsean committed:
compile

realbigsean committed:
cargo fix

realbigsean committed:
cargo fmt

realbigsean committed:
cleanup

realbigsean committed:
fix RPC limit add blob signing domain

realbigsean committed:
compressed eip4844 genesis

pawanjay176 committed:
Fix genesis.ssz.zip

pawanjay176 committed:
Add gossip verification stub

pawanjay176 committed:
Add more gossip verification functions for blobs

pawanjay176 committed:
Minor fixes

pawanjay176 committed:
Reprocess blob sidecar messages

pawanjay176 committed:
Add todos

realbigsean committed:
blob production

realbigsean committed:
cargo fix

pawanjay176 committed:
Add more gossip verification conditions

ethDreamer committed:
Added Capella Data Structures to consensus/types (#3637)

* Ran Cargo fmt

* Added Capella Data Structures to consensus/types

ethDreamer committed:
Fixed some stuff in state processing (#3640)

ethDreamer committed:
Fixed a ton of state_processing stuff (#3642)

FIXME's:
 * consensus/fork_choice/src/fork_choice.rs
 * consensus/state_processing/src/per_epoch_processing/capella.rs
 * consensus/types/src/execution_payload_header.rs
 
TODO's:
 * consensus/state_processing/src/per_epoch_processing/capella/partial_withdrawals.rs
 * consensus/state_processing/src/per_epoch_processing/capella/full_withdrawals.rs

realbigsean committed:
Capella eip 4844 cleanup (#3652)

* add capella gossip boiler plate

* get everything compiling

Co-authored-by: realbigsean <sean@sigmaprime.io
Co-authored-by: Mark Mackey <mark@sigmaprime.io>

* small cleanup

* small cleanup

* cargo fix + some test cleanup

* improve block production

* add fixme for potential panic

Co-authored-by: Mark Mackey <mark@sigmaprime.io>

ethDreamer committed:
Added Capella Epoch Processing Logic (#3666)

realbigsean committed:
clean up types

realbigsean committed:
48 byte array serde

pawanjay176 committed:
Couple blocks and blobs in gossip (#3670)

* Revert "Add more gossip verification conditions"

This reverts commit 1430b561c37adb44d5705005de6bf633deb8c16d.

* Revert "Add todos"

This reverts commit 91efb9d4c780b55025c3793a67bd9dacc1b2c924.

* Revert "Reprocess blob sidecar messages"

This reverts commit 21bf3d37cdce46632cfa4e3f5abb194f172c6851.

* Add the coupled topic

* Decode SignedBeaconBlockAndBlobsSidecar correctly

* Process Block and Blobs in beacon processor

* Remove extra blob publishing logic from vc

* Remove blob signing in vc

* Ugly hack to compile

realbigsean committed:
Block processing eip4844 (#3673)

* add eip4844 block processing

* fix blob processing code

* consensus logic fixes and cleanup

* use safe arith

realbigsean committed:
merge with unstable

realbigsean committed:
merge with unstable fixes

realbigsean committed:
Cleanup payload types (#3675)

* Add transparent support

* Add `Config` struct

* Deprecate `enum_behaviour`

* Partially remove enum_behaviour from project

* Revert "Partially remove enum_behaviour from project"

This reverts commit 46ffb7fe77622cf420f7ba2fccf432c0050535d6.

* Revert "Deprecate `enum_behaviour`"

This reverts commit 89b64a6f53d0f68685be88d5b60d39799d9933b5.

* Add `struct_behaviour`

* Tidy

* Move tests into `ssz_derive`

* Bump ssz derive

* Fix comment

* newtype transaparent ssz

* use ssz transparent and create macros for  per fork implementations

* use superstruct map macros

Co-authored-by: Paul Hauner <paul@paulhauner.com>

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into eip4844

realbigsean committed:
Feature gate withdrawals (#3684)

* start feature gating

* feature gate withdrawals

jimmygchen committed:
Fix compilation error (#3692)

realbigsean committed:
fix topic name

ethDreamer committed:
Updated for queueless withdrawals spec

ethDreamer committed:
Fixed compiling with withdrawals enabled

ethDreamer committed:
Added stuff that NEEDS IMPLEMENTING

ethDreamer committed:
BeaconState field renamed

ethDreamer committed:
Forgot one feature guard

ethDreamer committed:
Added process_withdrawals

michaelsproul committed:
Fixes to make EF Capella tests pass (#3719)

* Fixes to make EF Capella tests pass

* Clippy for state_processing

michaelsproul committed:
Fix BlocksByRoot response types (#3743)

ethDreamer committed:
Massive Update to Engine API (#3740)

* Massive Update to Engine API

* Update beacon_node/execution_layer/src/engine_api/json_structures.rs

Co-authored-by: Michael Sproul <micsproul@gmail.com>

* Update beacon_node/execution_layer/src/engine_api/json_structures.rs

Co-authored-by: Michael Sproul <micsproul@gmail.com>

* Update beacon_node/beacon_chain/src/execution_payload.rs

Co-authored-by: realbigsean <seananderson33@GMAIL.com>

* Update beacon_node/execution_layer/src/engine_api.rs

Co-authored-by: realbigsean <seananderson33@GMAIL.com>

Co-authored-by: Michael Sproul <micsproul@gmail.com>
Co-authored-by: realbigsean <seananderson33@GMAIL.com>

realbigsean committed:
- fix pre-merge block production (#3746)

- return `None` on pre-4844 blob requests

ethDreamer committed:
Stuuupid camelCase (#3748)

michaelsproul committed:
Two Capella bugfixes (#3749)

* Two Capella bugfixes

* fix payload default check in fork choice

* Revert "fix payload default check in fork choice"

This reverts commit e56fefbd05811526af4499711045275db366aa09.

Co-authored-by: realbigsean <sean@sigmaprime.io>

realbigsean committed:
Rename excess blobs and update 4844 json RPC serialization/deserialization (#3745)

* rename excess blobs and fix json serialization/deserialization

* remove coments

michaelsproul committed:
Op pool and gossip for BLS to execution changes (#3726)

ethDreamer committed:
Remove withdrawals guard for PayloadAttributesV2

ethDreamer committed:
Fixed some BeaconChain Tests

ethDreamer committed:
Fixed Payload Deserialization in DB (#3758)

ethDreamer committed:
Refactored Execution Layer & Fixed Some Tests

ethDreamer committed:
Merged with `unstable`

michaelsproul committed:
Merge pull request #3762 from ethDreamer/capella_eip4844_merge_unstable

merge `capella` with `unstable`

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/sigp/lighthouse?pullRequest=3763) <br/>All committers have signed the CLA.

ethDreamer committed:
Fixed Compiler Warnings & Failing Tests (#3771)

ethDreamer committed:
Merge 'upstream/unstable' into capella (#3773)

* Add API endpoint to count statuses of all validators (#3756)
* Delete DB schema migrations for v11 and earlier (#3761)

Co-authored-by: Mac L <mjladson@pm.me>
Co-authored-by: Michael Sproul <michael@sigmaprime.io>

ethDreamer committed:
Fixed moar tests (#3774)

jtraglia committed:
Fix some capella nits (#3782)

michaelsproul committed:
Fix `Withdrawal` serialisation and check address change fork (#3789)

* Disallow address changes before Capella

* Quote u64s in Withdrawal serialisation

ethDreamer committed:
Fixed Clippy Complaints & Some Failing Tests (#3791)

* Fixed Clippy Complaints & Some Failing Tests
* Update Dockerfile to Rust-1.65
* EF test file renamed
* Touch up comments based on feedback

ethDreamer committed:
Fixed Payload Reconstruction Bug (#3796)

michaelsproul committed:
Use JsonPayload for payload reconstruction (#3797)

michaelsproul committed:
Batch API for address changes (#3798)

michaelsproul committed:
Merge remote-tracking branch 'origin/unstable' into capella-update

michaelsproul committed:
Fix Clippy

michaelsproul committed:
Publish capella images on push (#3803)

michaelsproul committed:
Enable withdrawals features in Capella docker images (#3805)

ethDreamer committed:
Merge branch 'unstable' into capella

ethDreamer committed:
Merge pull request #3809 from ethDreamer/capella

Merge branch `unstable` into `capella`

michaelsproul committed:
Bounded withdrawals and spec v1.3.0-alpha.2 (#3802)

ethDreamer committed:
Make engine_getPayloadV2 accept local block value

ethDreamer committed:
Merge pull request #3817 from ethDreamer/capella

Make engine_getPayloadV2 accept local block value

ethDreamer committed:
Removed `withdrawals` feature flag

ethDreamer committed:
Update consensus/state_processing/src/upgrade/eip4844.rs

Co-authored-by: realbigsean <seananderson33@GMAIL.com>

ethDreamer committed:
Update consensus/state_processing/src/upgrade/eip4844.rs

Co-authored-by: realbigsean <seananderson33@GMAIL.com>

ethDreamer committed:
Merge pull request #3821 from ethDreamer/capella

Removed `withdrawals` feature flag

ethDreamer committed:
Fixed spec serialization bug

ethDreamer committed:
Merge pull request #3826 from ethDreamer/capella

Fixed spec serialization bug

ethDreamer committed:
Feature Guard V2 Engine API Methods

ethDreamer committed:
Fixed Some Tests

ethDreamer committed:
Merge pull request #3838 from ethDreamer/capella

Fixed Some Tests

ethDreamer committed:
merge upstream/unstable

ethDreamer committed:
Fix clippy complaints

michaelsproul committed:
Merge pull request #3842 from ethDreamer/capella

Merge with `unstable` and fix clippy

realbigsean committed:
cleanup

realbigsean committed:
Update beacon_node/execution_layer/src/engine_api/json_structures.rs

ethDreamer committed:
Update Execution Layer Tests for Capella

ethDreamer committed:
Fixed Operation Pool Tests

ethDreamer committed:
Fix EF Tests

ethDreamer committed:
Fixing Moar Failing Tests

ethDreamer committed:
Merge pull request #3847 from ethDreamer/capella

Update Execution Layer Tests for Capella

ethDreamer committed:
Merge pull request #3845 from realbigsean/capella-cleanup

Capella cleanup

ethDreamer committed:
Isolate withdrawals-processing Feature (#3854)

ethDreamer committed:
Added bls_to_execution_changes to PersistedOpPool (#3857)

* Added bls_to_execution_changes to PersistedOpPool

michaelsproul committed:
Bump MSRV to 1.65 (#3860)

realbigsean committed:
add historical summaries (#3865)

* add historical summaries

* fix tree hash caching, disable the sanity slots test with fake crypto

* add ssz static HistoricalSummary

* only store historical summaries after capella

* Teach `UpdatePattern` about Capella

* Tidy EF tests

* Clippy

Co-authored-by: Michael Sproul <michael@sigmaprime.io>

ethDreamer committed:
Remove `withdrawals-processing` feature (#3864)

* Use spec to Determine Supported Engine APIs

* Remove `withdrawals-processing` feature

* Fixed Tests

* Missed Some Spots

* Fixed Another Test

* Stupid Clippy

michaelsproul committed:
Fix Arbitrary implementations (#3867)

* Fix Arbitrary implementations

* Remove remaining vestiges of arbitrary-fuzz

* Remove FIXME

* Clippy

michaelsproul committed:
Merge remote-tracking branch 'origin/unstable' into capella

Fixing the conflicts involved patching up some of the `block_hash` verification,
the rest will be done as part of https://github.com/sigp/lighthouse/issues/3870

michaelsproul committed:
Merge pull request #3871 from michaelsproul/capella-v3.4.0

Back-merge unstable to Capella

michaelsproul committed:
Fix some beacon_chain tests

realbigsean committed:
Merge pull request #3874 from michaelsproul/beacon_chain_tests

Fix some beacon_chain tests

michaelsproul committed:
Verify blockHash with withdrawals

ethDreamer committed:
Sign BlsToExecutionChange w/ GENESIS_FORK_VERSION

ethDreamer committed:
Don't Penalize Early `bls_to_execution_change`

realbigsean committed:
Update gossip_methods.rs

realbigsean committed:
Merge pull request #3880 from ethDreamer/capella

Sign BlsToExecutionChange w/ GENESIS_FORK_VERSION

realbigsean committed:
bump ef-tests

realbigsean committed:
Merge pull request #3877 from michaelsproul/withdrawals-block-hash

Verify blockHash with withdrawals

realbigsean committed:
intentionally skip `LightClientHeader` ssz static tests

realbigsean committed:
Merge pull request #3881 from realbigsean/bump-ef-tests

bump ef-tests

ethDreamer committed:
CL-EL withdrawals harmonization using Gwei units (#3884)

ethDreamer committed:
Update engine_api to Latest spec (#3893)

* Update engine_api to Latest spec

* Small Test Fix

* Fix Test Deserialization Issue

michaelsproul committed:
Merge remote-tracking branch 'origin/unstable' into capella

michaelsproul committed:
Import BLS to execution changes before Capella (#3892)

* Import BLS to execution changes before Capella

* Test for BLS to execution change HTTP API

* Pack BLS to execution changes in LIFO order

* Remove unused var

* Clippy

ethDreamer committed:
Use eth1_withdrawal_credentials in Test States (#3898)

* Use eth1_withdrawal_credential in Some Test States

* Update beacon_node/genesis/src/interop.rs

Co-authored-by: Michael Sproul <micsproul@gmail.com>

* Update beacon_node/genesis/src/interop.rs

Co-authored-by: Michael Sproul <micsproul@gmail.com>

* Increase validator sizes

* Pick next sync committee message

Co-authored-by: Michael Sproul <micsproul@gmail.com>
Co-authored-by: Paul Hauner <paul@paulhauner.com>

michaelsproul committed:
Merge remote-tracking branch 'origin/unstable' into capella

michaelsproul committed:
Update sync rewards API for abstract exec payload

michaelsproul committed:
Fix the new BLS to execution change test

michaelsproul committed:
Update another test broken by the shuffling change

paulhauner committed:
Merge pull request #3914 from michaelsproul/capella-fix-tests

Fix the new BLS to execution change test

ethDreamer committed:
exchangeCapabilities & Capella Readiness Logging (#3918)

* Undo Passing Spec to Engine API

* Utilize engine_exchangeCapabilities

* Add Logging to Indicate Capella Readiness

* Add exchangeCapabilities to mock_execution_layer

* Send Nested Array for engine_exchangeCapabilities

* Use Mutex Instead of RwLock for EngineCapabilities

* Improve Locking to Avoid Deadlock

* Prettier logic for get_engine_capabilities

* Improve Comments

* Update beacon_node/beacon_chain/src/capella_readiness.rs

Co-authored-by: Michael Sproul <micsproul@gmail.com>

* Update beacon_node/beacon_chain/src/capella_readiness.rs

Co-authored-by: Michael Sproul <micsproul@gmail.com>

* Update beacon_node/beacon_chain/src/capella_readiness.rs

Co-authored-by: Michael Sproul <micsproul@gmail.com>

* Update beacon_node/beacon_chain/src/capella_readiness.rs

Co-authored-by: Michael Sproul <micsproul@gmail.com>

* Update beacon_node/beacon_chain/src/capella_readiness.rs

Co-authored-by: Michael Sproul <micsproul@gmail.com>

* Update beacon_node/client/src/notifier.rs

Co-authored-by: Michael Sproul <micsproul@gmail.com>

* Update beacon_node/execution_layer/src/engine_api/http.rs

Co-authored-by: Michael Sproul <micsproul@gmail.com>

* Addressed Michael's Comments

---------

Co-authored-by: Michael Sproul <micsproul@gmail.com>

ethDreamer committed:
Merge branch 'upstream/unstable' into capella

michaelsproul committed:
Merge pull request #3933 from ethDreamer/capella

Merge `unstable` into `capella`

ethDreamer committed:
Use Local Payload if More Profitable than Builder (#3934)

* Use Local Payload if More Profitable than Builder

* Rename clone -> clone_from_ref

* Minimize Clones of GetPayloadResponse

* Cleanup & Fix Tests

* Added Tests for Payload Choice by Profit

* Fix Outdated Comments

ethDreamer committed:
Don't Reject all Builder Bids After Capella (#3940)

* Fix bug in Builder API Post-Capella

* Fix Clippy Complaints

michaelsproul committed:
Remove unused `u256_hex_be_opt` (#3942)

paulhauner committed:
Broadcast address changes at Capella (#3919)

* Add first efforts at broadcast

* Tidy

* Move broadcast code to client

* Progress with broadcast impl

* Rename to address change

* Fix compile errors

* Use `while` loop

* Tidy

* Flip broadcast condition

* Switch to forgetting individual indices

* Always broadcast when the node starts

* Refactor into two functions

* Add testing

* Add another test

* Tidy, add more testing

* Tidy

* Add test, rename enum

* Rename enum again

* Tidy

* Break loop early

* Add V15 schema migration

* Bump schema version

* Progress with migration

* Update beacon_node/client/src/address_change_broadcast.rs

Co-authored-by: Michael Sproul <micsproul@gmail.com>

* Fix typo in function name

---------

Co-authored-by: Michael Sproul <micsproul@gmail.com>

ethDreamer committed:
Properly Deserialize ForkVersionedResponses (#3944)

* Move ForkVersionedResponse to consensus/types

* Properly Deserialize ForkVersionedResponses

* Elide Types in from_value Calls

* Added Tests for ForkVersionedResponse Deserialize

* Address Sean's Comments & Make Less Restrictive

* Utilize `map_fork_name!`

ethDreamer committed:
Update Mock Builder for Post-Capella Tests (#3958)

* Update Mock Builder for Post-Capella Tests

* Add _mut Suffix to BidStuff Functions

* Fix Setting Gas Limit

michaelsproul committed:
Merge remote-tracking branch 'origin/unstable' into capella-merge

michaelsproul committed:
Placeholder for BlobsByRange outbound rate limit

michaelsproul committed:
Update block rewards API for Capella

michaelsproul committed:
Merge pull request #3973 from michaelsproul/capella-merge

Update Capella to latest `unstable`

michaelsproul committed:
Quote Capella BeaconState fields (#3967)

michaelsproul committed:
Simplify payload traits and reduce cloning (#3976)

* Simplify payload traits and reduce cloning

* Fix self limiter

paulhauner commented:
I'm flagging this as ready for review! I think it's still OK to merge things in here, but I'm hoping we can get this merged early next week.

paulhauner commented on beacon_node/execution_layer/src/engine_api/http.rs:
```diff
@@ -519,10 +556,47 @@ pub mod deposit_methods {
+
+    pub fn engine_capabilities(&self) -> &EngineCapabilities {
+        &self.engine_capabilities
+    }
```
```suggestion
    pub fn engine_capabilities(&self) -> EngineCapabilities {
        self.engine_capabilities
    }
```

Super minor nit, but it makes a future suggestion slightly cleaner. `EngineCapabilities` is tiny and `Copy`, currently a reference/pointer is likely larger than the actual struct.

paulhauner commented on beacon_node/execution_layer/src/engine_api/http.rs:
```diff
@@ -736,6 +932,122 @@ impl HttpJsonRpc {
+        } else {
+            // here entry is guaranteed to exist so unwrap() is safe
+            Ok(*lock.as_ref().unwrap().engine_capabilities())
+        }
```
```suggestion
        if let Some(lock) = lock.as_ref().filter(|entry| !entry.older_than(age_limit)) {
            Ok(lock.engine_capabilities())
        } else {
            let engine_capabilities = self.exchange_capabilities().await?;
            *lock = Some(CapabilitiesCacheEntry::new(engine_capabilities));
            Ok(engine_capabilities)
        }
```

Removes the `unwrap()` and is still easy to read, IMO.

paulhauner commented on beacon_node/execution_layer/src/engine_api/http.rs:
```diff
@@ -681,31 +784,103 @@ impl HttpJsonRpc {
-        Ok(response.into())
+        Ok(GetPayloadResponse::Merge(GetPayloadResponseMerge {
+            execution_payload: payload_v1.into(),
+            // Have to guess zero here as we don't know the value
```
```suggestion
            // Set the V1 payload values from the EE to be zero. This simulates
            // the pre-block-value functionality of always choosing the builder
            // block.
```

Very minor nit, mostly represents that I took the time to go and double-check this.

paulhauner commented on beacon_node/execution_layer/src/engines.rs:
```diff
@@ -1,22 +1,26 @@
 /// The number of payload IDs that will be stored for each `Engine`.
 ///
 /// Since the size of each value is small (~100 bytes) a large number is used for safety.
+/// FIXME: check this assumption now that the key includes entire payload attributes which now includes withdrawals
```
```suggestion
/// Since the size of each value is small (~800 bytes) a large number is used for safety.
```

I think we can stick with a cache size of `512`. Here's my working:

- hash: `32`
- attributes:
  - timestamp: `8`
  - prev_randao: `32`
  - suggested_fee_recip: `20`
  - withdrawals: `44 * 16 = 704`
- payload_id: `8`

The size of each entry is `32+8+32+20+704+8 = 804` bytes.

The total cache would be `512 * 804 ~= 411KB` , which I think is very reasonable.

paulhauner commented on beacon_node/execution_layer/src/engines.rs:
```diff
@@ -273,7 +276,7 @@ impl Engine {
                 let mut state = self.state.write().await;
                 state.update(EngineStateInternal::AuthFailed);
-                **state
+                (**state, CapabilitiesCacheAction::None)
```
```suggestion
                (**state, CapabilitiesCacheAction::Clear)
```

I would expect to get an auth error in two cases:

1. We are connecting to the EE for the first time and a `Clear` can't hurt.
1. The EE has changed and it would be good to `Clear` and get fresh capabilities.

It's probably the case that in (2) we'd get an `Offline` first, but using `Clear` for auth errors still seems to make sense to me.

paulhauner reviewed (changes requested):
I just finished reviewing the `execution_layer`. It looks great, I only have really minor suggestions.

I believe that all of the suggestions here will compile if they're accepted and committed via the Github UI. Anything that I couldn't get to work via suggestions I added here: https://github.com/sigp/lighthouse/pull/3983

michaelsproul committed:
Execution engine suggestions from code review

Co-authored-by: Paul Hauner <paul@paulhauner.com>

paulhauner committed:
Suggestions for Capella `execution_layer` (#3983)

* Restrict Engine::request to FnOnce

* Use `Into::into`

* Impl IntoIterator for VariableList

* Use Instant rather than SystemTime

michaelsproul committed:
Merge remote-tracking branch 'origin/staging' into capella-update

michaelsproul committed:
Merge pull request #3981 from michaelsproul/capella-update

Update `capella` to `unstable`

paulhauner committed:
Add capella fork epoch (#3997)

paulhauner commented:
I just finished reviewing `beacon_chain`, my suggestions are at #3999. It's looking good!

michaelsproul committed:
Fix Capella schema downgrades (#4004)

paulhauner committed:
Remove "eip4844" network (#4008)

paulhauner committed:
Suggestions for Capella `beacon_chain` (#3999)

* Remove CapellaReadiness::NotSynced

Some EEs have a habit of flipping between synced/not-synced, which causes some
spurious "Not read for the merge" messages back before the merge. For the
merge, if the EE wasn't synced the CE simple wouldn't go through the transition
(due to optimistic sync stuff). However, we don't have that hard requirement
for Capella; the CE will go through the fork and just wait for the EE to catch
up. I think that removing `NotSynced` here will avoid false-positives on the
"Not ready logs..". We'll be creating other WARN/ERRO logs if the EE isn't
synced, anyway.

* Change some Capella readiness logging

There's two changes here:

1. Shorten the log messages, for readability.
2. Change the hints.

Connecting a Capella-ready LH to a non-Capella-ready EE gives this log:

```
WARN Not ready for Capella                   info: The execution endpoint does not appear to support the required engine api methods for Capella: Required Methods Unsupported: engine_getPayloadV2 engine_forkchoiceUpdatedV2 engine_newPayloadV2, service: slot_notifier
```

This variant of error doesn't get a "try updating" style hint, when it's the
one that needs it. This is because we detect the method-not-found reponse from
the EE and return default capabilities, rather than indicating that the request
fails. I think it's fair to say that an EE upgrade is required whenever it
doesn't provide the required methods.

I changed the `ExchangeCapabilitiesFailed` message since that can only happen
when the EE fails to respond with anything other than success or not-found.

paulhauner committed:
Capella consensus review (#4012)

* Add extra encoding/decoding tests

* Remove TODO

The method LGTM

* Remove `FreeAttestation`

This is an ancient relic, I'm surprised it still existed!

* Add paranoid check for eip4844 code

This is not technically necessary, but I think it's nice to be explicit about
EIP4844 consensus code for the time being.

* Reduce big-O complexity of address change pruning

I'm not sure this is *actually* useful, but it might come in handy if we see a
ton of address changes at the fork boundary. I know the devops team have been
testing with ~100k changes, so maybe this will help in that case.

* Revert "Reduce big-O complexity of address change pruning"

This reverts commit e7d93e6cc7cf1b92dd5a9e1966ce47d4078121eb.

paulhauner committed:
Revert Sepolia genesis change (#4013)

paulhauner committed:
Allow for withdrawals in max block size (#4011)

* Allow for withdrawals in max block size

* Ensure payload size is counted

paulhauner committed:
Fix post-Bellatrix checkpoint sync (#4014)

* Recognise execution in post-merge blocks

* Remove `.body()`

* Fix typo

* Use `is_default_with_empty_roots`.

paulhauner committed:
Modify some Capella comments (#4015)

* Modify comment to only include 4844

Capella only modifies per epoch processing by adding
`process_historical_summaries_update`, which does not change the realization of
justification or finality.

Whilst 4844 does not currently modify realization, the spec is not yet final
enough to say that it never will.

* Clarify address change verification comment

The verification of the address change doesn't really have anything to do with
the current epoch. I think this was just a copy-paste from a function like
`verify_exit`.

paulhauner reviewed (approved):
Approved! 🎉 🎉 

Congrats to all involved, this looks great! As you can see in the history, I made some suggestions via a bunch of PRs. However I don't think any of them were really significant, we probably could have gone to production as it was.

Seeing as we're planning to merge this into `unstable` via a merge commit (rather than squash merge via bors), I think it makes sense to wait and see that CI finishes before merging.

divagant-martian commented on beacon_node/network/src/beacon_processor/mod.rs:
```diff
@@ -1408,7 +1517,7 @@ impl<T: BeaconChainTypes> BeaconProcessor<T> {
                 );
                 metrics::set_gauge(
-                    &metrics::BEACON_PROCESSOR_RPC_BLOCK_QUEUE_TOTAL,
+                    &metrics::BEACON_PROCESSOR_RPC_BLOB_QUEUE_TOTAL,
```
this change should be reverted

paulhauner replied:
Good catch, I fixed this in #4020

pawanjay176 committed:
Remove BeaconBlockAndBlobsSidecar from core topics (#4016)

paulhauner committed:
Fix metric (#4020)

michaelsproul commented:
Leaving the `capella` branch around for people using old instructions to build from source. We can delete it in a few days.
